runtime/compiler/optimizer/TreeLowering.cpp

```diff
@@ -719,3 +756,719 @@ TR::TreeLowering::lowerArrayStoreCHK(TR::Node *node, TR::TreeTop *tt)
+   if (storeArrayElementNode->getOpCodeValue() == TR::aRegStore)
+      {
+      TR::Node *blkDeps = arrayElementLoadBlock->getExit()->getNode()->getFirstChild();
+
+      TR::Node *depNode = TR::Node::create(TR::PassThrough, 1, storeArrayElementNode->getChild(0));
+      depNode->setGlobalRegisterNumber(storeArrayElementNode->getGlobalRegisterNumber());
+      blkDeps->addChildren(&depNode, 1);
+      }
+
+
+   ///////////////////////////////////////
+   // 11. Set up the edges between the blocks
+   ifNode->setBranchDestination(helperCallBlock->getEntry());
+
+   cfg->addEdge(originalBlock, helperCallBlock);
+
+   cfg->removeEdge(arrayElementLoadBlock, helperCallBlock);
+   cfg->addEdge(arrayElementLoadBlock, blockAfterHelperCall);
+
+   // Force nodeIter to first TreeTop of next block so that
+   // moving callBlock won't cause problems while iterating
+   while (nodeIter.currentTree() != blockAfterHelperCall->getEntry())
+       ++nodeIter;
+
+   // Move helper call to the end of the tree list
+   arrayElementLoadBlock->getExit()->join(blockAfterHelperCall->getEntry());
+
+   TR::TreeTop *lastTreeTop = comp->getMethodSymbol()->getLastTreeTop();
+   lastTreeTop->insertTreeTopsAfterMe(helperCallBlock->getEntry(), helperCallBlock->getExit());
+
+   // Add Goto from the helper call to the merge block
+   TR::Node *gotoAfterHelperCallNode = TR::Node::create(helperCallBlock->getExit()->getNode(), TR::Goto, 0, blockAfterHelperCall->getEntry());
+   helperCallBlock->append(TR::TreeTop::create(comp, gotoAfterHelperCallNode));
+
+   if (helperCallBlock->getExit()->getNode()->getNumChildren() > 0)
+      {
+      TR::Node *deps = helperCallBlock->getExit()->getNode()->getChild(0);
+      helperCallBlock->getExit()->getNode()->setNumChildren(0);
+      deps->decReferenceCount();
+      gotoAfterHelperCallNode->addChildren(&deps, 1);
+      }
+   }
+
+/*
+ * lowerStoreArrayElement transforms the block that contains the jitStoreFlattenableArrayElement helper call into three blocks:
+ *   1. The merge block that contains the tree tops after the helper call
+ *   2. The helper call block that contains the helper call and is moved to the end of the tree top list
+ *   3. The new non-VT array store block which is an extended block of the original block
+ *
+ *      originalBlock ----------+
+ *      arrayElementStoreBlock  |
+ *               |              |
+ *               |              v
+ *               |       helperCallBlock
+ *               |              |
+ *               v              v
+ *             blockAfterHelperCall
+ *
+ +-------------------------------------------+
+ |treetop                                    |
+ |   acall  jitStoreFlattenableArrayElement  |
+ |      aload <value>                        |
+ |      iload <index>                        |
+ |      aload <arrayAddress>                 |
+ |ttAfterHelperCall                          |
+ +-------------------------------------------+
+               |              |
```

Comment:
This might sound silly, but I was confused for a couple of seconds by this pair of arrows.  I quickly realized that they're indicating that a transformation has occurred, but it might be good to say that in words by labellng things "Before" and "After" or something like that.